// arui-feather/message/message.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _dec, _dec2, _class, _class2, _temp; /* This Source Code Form is subject to the terms of the Mozilla Public
                                          * License, v. 2.0. If a copy of the MPL was not distributed with this
                                          * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _popup = require('../popup/popup');

var _popup2 = _interopRequireDefault(_popup);

var _cn = require('../cn');

var _cn2 = _interopRequireDefault(_cn);

var _performance = require('../performance');

var _performance2 = _interopRequireDefault(_performance);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

/**
 * Компонент сообщения.
 */
var Message = (_dec = (0, _cn2.default)('message'), _dec2 = (0, _performance2.default)(), _dec(_class = _dec2(_class = (_temp = _class2 = function (_React$Component) {
    _inherits(Message, _React$Component);

    function Message() {
        _classCallCheck(this, Message);

        return _possibleConstructorReturn(this, (Message.__proto__ || Object.getPrototypeOf(Message)).apply(this, arguments));
    }

    _createClass(Message, [{
        key: 'componentDidMount',
        value: function componentDidMount() {
            if (this.popup && this.popupTarget) {
                this.popup.setTarget(this.popupTarget);
            }
        }
    }, {
        key: 'render',
        value: function render(cn) {
            return this.props.type === 'text' ? this.renderTextMessage(cn) : this.renderPopupMessage(cn);
        }
    }, {
        key: 'renderTextMessage',
        value: function renderTextMessage(cn) {
            return _react2.default.createElement(
                'div',
                {
                    className: cn({
                        type: 'text',
                        showed: this.props.visible
                    }) + ' message__control'
                },
                this.props.children
            );
        }
    }, {
        key: 'renderPopupMessage',
        value: function renderPopupMessage(cn) {
            var _this2 = this;

            return _react2.default.createElement(
                'div',
                null,
                _react2.default.createElement(
                    'div',
                    {
                        className: cn({ type: 'popup' }),
                        ref: function ref(popupTarget) {
                            _this2.popupTarget = popupTarget;
                        }
                    },
                    _react2.default.createElement(
                        _popup2.default,
                        {
                            className: cn('control'),
                            ref: function ref(popup) {
                                _this2.popup = popup;
                            },
                            visible: this.props.visible
                        },
                        this.props.children
                    )
                )
            );
        }
    }]);

    return Message;
}(_react2.default.Component), _class2.propTypes = {
    /** Тип компонента */
    type: _propTypes2.default.oneOf(['text', 'popup']),
    /** Управление видимостью компонента */
    visible: _propTypes2.default.bool,
    /** Дочерние элементы `Message` */
    children: _propTypes2.default.oneOfType([_propTypes2.default.arrayOf(_propTypes2.default.node), _propTypes2.default.node]),
    /** Тема компонента */
    theme: _propTypes2.default.oneOf(['alfa-on-color', 'alfa-on-white']),
    /** Дополнительный класс */
    className: _propTypes2.default.oneOfType([_propTypes2.default.func, _propTypes2.default.string])
}, _class2.defaultProps = {
    type: 'text'
}, _temp)) || _class) || _class);
exports.default = Message;
